crates/bevy_utils/src/futures.rs

```diff
@@ -5,16 +5,22 @@ use std::{
     task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
 };
 
-/// Consumes the future, polls it once, and immediately returns the output
+/// Consumes a future, polls it once, and immediately returns the output
 /// or returns `None` if it wasn't ready yet.
 ///
 /// This will cancel the future if it's not ready.
 pub fn now_or_never<F: Future>(mut future: F) -> Option<F::Output> {
+    check_ready(&mut future)
+}
+
+/// Polls a future once, and returns the output if ready
+/// or returns `None` if it wasn't ready yet.
+pub fn check_ready<F: Future>(future: &mut F) -> Option<F::Output> {
     let noop_waker = noop_waker();
     let mut cx = Context::from_waker(&noop_waker);
 
     // SAFETY: `future` is not moved and the original value is shadowed
-    let future = unsafe { Pin::new_unchecked(&mut future) };
+    let future = unsafe { Pin::new_unchecked(future) };
```

Comment:
pretty sure this is unsafe

> A value, once pinned, must remain pinned until it is dropped (unless its type implements Unpin). Because Pin<&mut T> does not own the value, dropping the Pin will not drop the value and will not end the pinning contract. So moving the value after dropping the Pin<&mut T> is still a violation of the API contract.

originally we owned the value so could be sure we would not move it, but that's not the case when we take `&mut F`. 

```rust
pub fn check_ready<F: Future + Unpin>(future: &mut F) -> Option<F::Output> {
    let noop_waker = noop_waker();
    let mut cx = Context::from_waker(&noop_waker);

    let future = Pin::new(future);

    match future.poll(&mut cx) {
        Poll::Ready(x) => Some(x),
        _ => None,
    }
}
```

works though since `Task` implements `Unpin` (and you need to revert the original now_or_never as well).